### DISCUSSION TITLE
Customizing error_handler

### DISCUSSION THREAD

**Discussion by colmexdev** [jeremyevans/forme#126] [General]
Hi there!

I'm also messing a bit with this gem inside a project and, so far, it's been quite straightforward. Nonetheless, I was surprised to see that error messages are automatically appended next to the corresponding input (which is great).

Then only thing I couldn't figure out and was wondering if it's indeed possible is how to customize generated tags for error messages (e.g. changing classes, tag  name or even show them as a `<ul>` instead of consecutive `<span>` tags.

Thank you very much in advance.

**Comment by jeremyevans** [comment]
Sure, it's possible to customize the error handler via the :error_handler option.  The default  inserts a span after the input, but you can customize the handling.  You can pass any callable object as the value of :error_handler.  It's probably best to review the implementation, as well as see the example of a custom :error_handler in the specs.  If you have additional questions after reviewing those, please let me know.

Code: https://github.com/jeremyevans/forme/blob/master/lib/forme/transformers/error_handler.rb
Spec: https://github.com/jeremyevans/forme/blob/master/spec/forme_spec.rb#L832

**Reply by colmexdev** [reply]
It's working now! Thank you very much for such concise info.

Out of curiosity, could there be a way to wrap only the error tags? What I'm thinking of is showing a list of every error tag per input in the form. But for what I can see, it is only possible to customize each error tag entry, not wrapping them up together aside of input and label.

**Reply by jeremyevans** [reply]
You could write an error handler that records each error instead of changing the output: 

```ruby
errors = {}
:error_handler=>proc{|t, i| errors[i] = i.opts[:error]; t}
```

And then later iterate over the errors to display them all together.   I suppose with some CSS, you could get them displaying above the form instead of below.

However, I think when I've done this (for forms for Sequel models), I've just iterated over `model_obj.errors.full_messages` and shown the errors before the form input.

**Reply by colmexdev** [reply]
Great! It's also working as expected!

Thank you very much @jeremyevans . Really appreciate your patience and ease of explanation.
